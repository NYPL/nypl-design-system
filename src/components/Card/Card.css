@import "../Heading/Heading.css";

@custom-media --card-break (max-width var(--breakpoint-medium));

.card {
  align-items: flex-start;
  border: 1px solid var(--ui-gray-light);
  border-radius: 5px;
  display: flex;
  flex-flow: column wrap;
  margin-bottom: var(--space-s);
  padding: var(--space) var(--space-l);

  &__heading,
  &__footer {
    flex: 100%;
  }

  &__columns {
    display: flex;
    flex-flow: column wrap;
    justify-content: space-between;
    width: 100%;

    @media (--card-break) {
      flex-flow: row wrap;
    }
  }

  &__columns + &__footer {
    margin-top: var(--space-xs);
  }

  &__image {
    align-items: center;
    background-color: var(--ui-gray-xlight);
    display: flex;
    flex: 0 0 150px;
    height: 150px;
    justify-content: center;
    margin-bottom: var(--space-s);

    .image {
      max-height: 100%;
      max-width: 100%;
    }

    @media (--card-break) {
      align-self: center;
      margin-bottom: 0;
      margin-right: var(--space-s);
      width: 150px;
    }
  }

  &__content {
    flex: 1 1;
    margin-bottom: var(--space-s);

    @media (--card-break) {
      margin-bottom: 0;
    }
  }

  &__ctas {
    flex: 0 1 20%;
  }

  &__content + &__ctas {
    margin-right: var(--space-s);
  }
}

/* Stories-only styles */
.sb-show-main {
  .var-docs {
    padding: 0 0 var(--space) 0;
  }

  .example-card {
    .card__content {
      > * {
        margin-bottom: var(--space-xxs);
      }
    }

    .book__callout {
      @add-mixin heading-xs;

      font-style: italic;
      font-weight: 300;
    }

    .card__ctas {
      align-items: flex-end;
      display: flex;
      flex-flow: row wrap;
      justify-content: center;
      text-align: center;

      .button {
        width: 100%;
      }

      > * {
        margin-bottom: var(--space-xxs);
      }
    }

    .book__filetype {
      margin-bottom: var(--space-s);
    }

    .book__description {
      font-style: italic;
    }
  }
}
