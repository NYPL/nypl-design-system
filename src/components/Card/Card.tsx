import * as React from "react";
import bem from "../../utils/bem";

interface CardProps {
  /** ClassName that appears in addition to "card" */
  className?: string;
  /** BlockName for use with BEM. See how to work with blockNames and BEM here: http://getbem.com/introduction/ */
  blockName?: string;
  /** Calls to Action for the card, such as "download" on a card for a book */
  ctas?: React.ReactNode;
  /** Footer that appears within the card but below the three columns within it */
  footer?: React.ReactNode;
  /** Heading that appears within the card but above the three columns within it */
  heading?: React.ReactNode;
  /** ID that other components can cross reference for accessibility purposes */
  id?: string;
  /** First column within the card */
  image?: React.ReactNode;
  /** Modifiers array for use with BEM. See how to work with modifiers and BEM here: http://getbem.com/introduction/ */
  modifiers?: string[];
}

export default function Card(props: React.PropsWithChildren<CardProps>) {
  const {
    blockName,
    children,
    className,
    ctas,
    footer,
    heading,
    id,
    image,
    modifiers = [],
  } = props;
  const baseClass = "card";

  return (
    <div className={bem(baseClass, modifiers, blockName, [className])} id={id}>
      {heading && (
        <div className={bem("heading", modifiers, baseClass)}>{heading}</div>
      )}
      <div className={bem("columns", modifiers, baseClass)}>
        {image && (
          <div className={bem("image", modifiers, baseClass)}>{image}</div>
        )}
        {children && (
          <div className={bem("content", modifiers, baseClass)}>{children}</div>
        )}
        {ctas && (
          <div className={bem("ctas", modifiers, baseClass)}>{ctas}</div>
        )}
      </div>
      {footer && (
        <div className={bem("footer", modifiers, baseClass)}>{footer}</div>
      )}
    </div>
  );
}
