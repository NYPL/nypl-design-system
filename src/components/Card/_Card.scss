.card {
  @include space-stack-s;

  align-items: flex-start;
  border: 1px solid var(--ui-gray-light);
  border-radius: 5px;
  display: flex;
  flex-flow: column wrap;
  padding: var(--space) var(--space-l);

  &__heading,
  &__footer {
    flex: 100%;
  }

  &__columns {
    display: flex;
    flex-flow: column wrap;
    justify-content: space-between;
    width: 100%;

    @include breakpoint($breakpoint-medium) {
      flex-flow: row wrap;
    }
  }

  &__columns + &__footer {
    margin-top: var(--space-xs);
  }

  &__image {
    @include space-stack-s;

    align-items: center;
    background-color: var(--ui-gray-xlight);
    display: flex;
    flex: 0 0 150px;
    height: 150px;
    justify-content: center;

    .image {
      max-height: 100%;
      max-width: 100%;
    }

    @include breakpoint($breakpoint-medium) {
      @include space-stack-none;
      @include space-inline-s;

      align-self: center;
      width: 150px;
    }
  }

  &__content {
    @include space-stack-s;

    flex: 1 1;

    @include breakpoint($breakpoint-medium) {
      @include space-inline-s;
      @include space-stack-none;
    }
  }

  &__ctas {
    flex: 0 1 20%;

    @include breakpoint($breakpoint-medium) {
      @include space-inline-s;
    }
  }

  &__content + &__ctas {
    @include space-inline-s;
  }
}

// Stories-only styles
.sb-show-main {
  .var-docs {
    padding: 0 0 var(--space) 0;
  }

  .example-card {
    .card__content {
      > * {
        @include space-stack-xxs;
      }
    }

    .book__callout {
      @include heading-xs;

      font-style: italic;
      font-weight: 300;
    }

    .card__ctas {
      align-items: flex-end;
      display: flex;
      flex-flow: row wrap;
      justify-content: center;
      text-align: center;

      .button {
        width: 100%;
      }

      > * {
        @include space-stack-xxs;
      }
    }

    .book__filetype {
      @include space-stack-s;
    }

    .book__description {
      font-style: italic;
    }
  }
}
