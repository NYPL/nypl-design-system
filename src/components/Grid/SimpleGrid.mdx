import { ArgTypes, Canvas, Description, Meta } from "@storybook/blocks";

import * as SimpleGridStories from "./SimpleGrid.stories";

<Meta of={SimpleGridStories} />

# SimpleGrid

| Component Version | DS Version |
| ----------------- | ---------- |
| Added             | `0.25.1`   |
| Latest            | `1.0.6`    |

## Table of Contents

- [Overview](#overview)
- [Component Props](#component-props)
- [Accessibility](#accessibility)
- [Other SimpleGrid Examples](#other-simplegrid-examples)

## Overview

<Description of={SimpleGridStories} />

The `SimpleGrid` component is used to render UI elements in a uniform grid
layout, with the column widths and grid spacing fixed throughout the grid.

The NYPL standards for the items per row in a grid is `3` for desktop, `2` for
tablet and `1` for mobile. By default, the `SimpleGrid` component uses these
standards and the `columns` prop is optional. If the `columns` prop is used, the
tablet breakpoint will be dropped and only the mobile breakpoint (1 item per
row) will be triggered.

## Component Props

<Canvas of={SimpleGridStories.WithControls} />

<ArgTypes of={SimpleGridStories.WithControls} />

## Accessibility

The CSS grid layout properties are used for the `SimpleGrid` component. We don't
recommend using property rules that change the visual order of elements on the
page that don't match with its DOM order. This is because a screenreader won't
pick up `SimpleGrid` CSS rules and will read the page in the expected DOM order
rather than the visual order.

Resources:

- [MDN CSS Grid Layout and Accessibility](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Grid_Layout/CSS_Grid_Layout_and_Accessibility)

## Other SimpleGrid Examples

### Icons

This example is rendering `Icon` components with the `columns` prop set to `5`.

<Canvas of={SimpleGridStories.IconExample} />

### Images

This example is rendering square `Image` components with the `columns` prop set
to `6`.

<Canvas of={SimpleGridStories.ImageExample} />

### Horizontal Cards

This example is rendering horizontal `Card` components with the `columns` prop
set to `1`.

<Canvas of={SimpleGridStories.CardExample} />
