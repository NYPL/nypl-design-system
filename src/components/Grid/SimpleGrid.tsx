import { chakra, SimpleGrid as ChakraSimpleGrid } from "@chakra-ui/react";
import React, { forwardRef } from "react";

export const gridGapsArray = [
  "grid.xxs",
  "grid.xs",
  "grid.s",
  "grid.m",
  "grid.l",
  "grid.xl",
  "grid.xxl",
] as const;
export type GridGaps = typeof gridGapsArray[number];

export interface SimpleGridProps {
  /** Additional class name. */
  className?: string;
  /** Optional numeric value to override the default column count; the default
   * column count is 3. */
  columns?: number;
  /** Optional gap size; if omitted, the default `large` (2rem / 32px) spacing
   * will be used; `IMPORTANT: for standard grid layouts, this prop should
   * not be used.` */
  gap?: GridGaps;
  /** ID that other components can cross reference for accessibility purposes */
  id?: string;
}

export const SimpleGrid: React.FC<any> = chakra(
  forwardRef<HTMLDivElement, React.PropsWithChildren<SimpleGridProps>>(
    (props, ref?) => {
      const {
        children,
        columns,
        className,
        gap = "grid.l",
        id,
        ...rest
      } = props;

      const responsiveCols = columns
        ? { base: 1, md: columns }
        : { base: 1, md: 2, lg: 3 };

      return (
        <ChakraSimpleGrid
          className={className}
          columns={responsiveCols}
          gap={gap}
          id={id}
          ref={ref}
          {...rest}
        >
          {children}
        </ChakraSimpleGrid>
      );
    }
  ),
  { shouldForwardProp: () => true }
);

export default SimpleGrid;
