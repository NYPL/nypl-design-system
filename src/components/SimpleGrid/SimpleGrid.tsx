import * as React from "react";
import { SimpleGrid as ChakraSimpleGrid } from "@chakra-ui/react";
import { GridGaps } from "./SimpleGridTypes";

import generateUUID from "../../helpers/generateUUID";

export interface SimpleGridProps {
  /** Additional class name. */
  className?: string;
  /** Optional numeric value to override the default column count; the default column count is 3 */
  columns?: number;
  /** Optional gap size; if omitted, the default `large` (2rem / 32px) spacing will be used; ```IMPORTANT: for general grid layout, this prop should not be used``` */
  gap?: GridGaps;
  /** ID that other components can cross reference for accessibility purposes */
  id?: string;
}

//  * Map the ButtonType to the Button Chakra theme variant object. If a wrong
//  * value is passed (typically in non-Typescript scenarios), then the default
//  * is the "primary" variant.
// const getVariant = (variant, collection) => {
//   // Used to map between enum values and Chakra variant options.
//   const variantMap = {};
//   for (const type in collection) {
//     variantMap[collection[type]] = collection[type];
//   }
//   return variantMap[variant] || collection.Default;
// };

function SimpleGrid(props: React.PropsWithChildren<SimpleGridProps>) {
  const { children, columns = 3, className, gap = GridGaps.Large, id } = props;

  // const variant = gap ? getVariant(gap, GridGaps) : null;
  // const styles = useStyleConfig("Grid", { variant });
  // console.log(`variant == ${variant}`);

  // const colConfig = `repeat(${columnCount}, 1fr)`;

  return (
    <ChakraSimpleGrid
      columns={{ base: 1, md: columns }}
      gap={gap}
      id={id || generateUUID()}
      className={className}
      {...props}
    >
      {children}
    </ChakraSimpleGrid>
  );
}

export default SimpleGrid;
