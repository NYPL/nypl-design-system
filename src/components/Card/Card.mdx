import { ArgTypes, Canvas, Description, Meta, Source } from "@storybook/blocks";

import * as CardStories from "./Card.stories";
import Link from "../Link/Link";
import { changelogData } from "./cardChangelogData";
import ComponentChangelogTable from "../../utils/ComponentChangelogTable";

<Meta of={CardStories} />

# Card

| Component Version | DS Version |
| ----------------- | ---------- |
| Added             | `0.24.0`   |
| Latest            | `3.1.2`    |

## Table of Contents

- {<Link href="#overview" target="_self">Overview</Link>}
- {<Link href="#component-props" target="_self">Component Props</Link>}
- {<Link href="#accessibility" target="_self">Accessibility</Link>}
- {<Link href="#image-position" target="_self">Image Position</Link>}
- {<Link href="#image-size" target="_self">Image Size</Link>}
- {<Link href="#custom-image-component" target="_self">Custom Image Component</Link>}
- {<Link href="#card-with-link-heading" target="_self">Card With Link Heading</Link>}
- {<Link href="#card-with-full-click-functionality" target="_self">Card With Full-Click Functionality</Link>}
- {<Link href="#card-with-right-side-cardactions" target="_self">Card with Right Side CardActions</Link>}
- {<Link href="#cards-in-a-grid" target="_self">Cards in a Grid</Link>}
- {<Link href="#cards-in-a-stack" target="_self">Cards in a Stack</Link>}
- {<Link href="#cards-without-images" target="_self">Cards Without Images</Link>}
- {<Link href="#changelog" target="_self">Changelog</Link>}

## Overview

<Description of={CardStories} />

The `Card` component is viewed as a contained, stand-alone element. It is highly
configurable and composable. It can be shown with or without an image and the
content area is composed using Card-specific child components: `CardHeading`,
`CardActions` and `CardContent`. These child components can be arranged in any
order and each child component may be used multiple times within the content
area. Although a `Card` is a self-contained unit, `Card` components are often
grouped together and displayed in some type of grid layout.

Note: all of the examples have images with the `isLazy` set to true.

### CardHeading

The `CardHeading` component mirrors the standard Reservoir Design System (DS)
`Heading` component and accepts the [same props](../?path=/story/components-typography-styles-heading--with-controls).

### CardActions

The `CardActions` component is used to display "call to action" (CTA) buttons
and links for the card. The DS `Button` and `Link` components should be passed
as children into the `CardActions` component.

Set the `isAlignedRightActions` prop to true in the `Card` component to align the
`CardActions` to the right of the the main content area. This is only applicable
in the row layout of the `Card` component.

### CardContent

The `CardContent` component should be used to display all content other than
headings and CTAs.

### CardImage

The `CardImage` component is _not_ exported and is used internally only in the
`Card` component. It is a convenience component that mirrors the standard DS
`Image` component. This means that the `CardImage` component accepts the same
props as the `Image` component, in addition to the `isAtEnd` prop.

Developers do _not_ need to implement this component. Instead, it will be
rendered in the `Card` component through its `imageProps` prop.

<Source
  code={`
<Card
  imageProps={{
    alt: "Alt text",
    aspectRatio: "twoByOne",
    caption: "caption text",
    credit: "credit text",
    isAtEnd: true,
    isLazy: true,
    size: "medium",
    src: "//loremflickr.com/400/200/new+york+public+library",
  }}
  {...otherCardProps}
>
  <CardHeading level="h3" id="heading-id">
    Heading Text
  </CardHeading>
  <CardContent>
    Praesent commodo cursus magna, vel scelerisque nisl consectetur et.
    Sed posuere consectetur est at lobortis. Cras justo odio, dapibus ac
    facilisis in, egestas eget quam.
  </CardContent>
</Card>
`}
  language="jsx"
/>

## Component Props

<Canvas of={CardStories.WithControls} />

<ArgTypes of={CardStories.WithControls} />

## Accessibility

### Headings

In general, headings from `h1` through `h6` must be in sequential order in the
DOM tree. Rendering an `h4` following an `h2` will throw an accessibility error.
Therefore, when adding `CardHeading` components, make sure to add them in
sequential order with the proper `level` prop.

<Source
  code={`
// Correct

<CardHeading level="h3" id="headingthree">
  Heading level three
</CardHeading>
<CardHeading level="h4" id="headingfour">
  Heading level four
</CardHeading>

// Incorrect

<CardHeading level="h2" id="headingtwo">
  Heading level Two
</CardHeading>
<CardHeading level="h4" id="headingfour">
  Heading level four
</CardHeading>
`} language="jsx" />

### Full-Click Functionality

Passing in a URL in the `mainActionLink` prop will make the entire `Card`
clickable. Other links in the `CardActions` component can still be accessed by
tabbing through and pressing "enter" or by clicking as you normally would on a
link.

## Image Position

By default, the image is always the first element within a `Card`. This means
the image is at the top of a `Card` with a column layout and on the left side of
a `Card` with a row layout. However, you can use the `imageProps.isAtEnd` boolean
prop to override the default placement and move the image to the last element
within a `Card`.

### Column Cards

<Canvas of={CardStories.ImagePositionColumnCards} />

### Row Cards

<Canvas of={CardStories.ImagePositionRowCards} />

## Image Size

By default, the image width is `100%` for a `Card` with a column layout and
`225px` for a `Card` with a row layout. However, you can use the `imageProps.size`
prop to override the default width.

### Column Cards

<Canvas of={CardStories.ImageSizeColumnCards} />

### Row Cards

<Canvas of={CardStories.ImageSizeRowCards} />

## Custom Image Component

Instead of passing a path for the `imageProps.src` prop, you can pass a custom
image component in using the `imageProps.component` prop.

<Canvas of={CardStories.CustomImageComponent} />

## Card with Link Heading

The `CardHeading` component works just like the DS `Heading` component. This
means that a URL can be passed in the `url` prop for `CardHeading` or a `Link`
component can be used as a child. Click on "Show code" in the example below to
see both patterns.

<Canvas of={CardStories.HeadingAsLink} />

## Card With Full-Click Functionality

To enable the full-click functionality in the `Card` component, pass a URL in
the `mainActionLink` prop. This will make the entire `Card` component clickable.
If the `Card` has action links or buttons, they can still be accessed by tabbing
through the `Card` and pressing `enter` or clicking like a regular link.

Internally, if multiple `CardHeading` components are passed, only the first one
will have the full-click link. This, however, does not affect the `Card` itself
from being having the full-click functionality.

<Canvas of={CardStories.FullClick} />

This example can be found in the [Turbine homepage](https://nypl-ds-test-app.vercel.app/).

<Canvas of={CardStories.CardFullClickTurbineExample} />

## Card With Full-Click Functionality and Tooltip

<Canvas of={CardStories.FullClickWithTooltip} />

## Card with Right Side CardActions

It's possible to set only the `CardActions` component on the right side of the
main content area of the `Card`. This is possible only in the `layout="row"`
layout and through the `Card` component's `isAlignedRightActions` prop which
must be set to `true`.

<Canvas of={CardStories.CardWithRightSideCardActions} />

## Cards in a Grid

<Canvas of={CardStories.GridExample} />

## Cards in a Stack

<Canvas of={CardStories.StackExample} />

## Cards Without Images

<Canvas of={CardStories.WithoutImages} />

## Changelog

<ComponentChangelogTable changelogData={changelogData} />
