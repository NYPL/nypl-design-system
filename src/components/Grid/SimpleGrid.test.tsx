import { render } from "@testing-library/react";
import { axe } from "jest-axe";
import * as React from "react";
import renderer from "react-test-renderer";

import Card, { CardHeading, CardContent } from "../Card/Card";
import SimpleGrid from "./SimpleGrid";

describe("Grid Accessibility", () => {
  it("passes axe accessibility test with children components", async () => {
    const { container } = render(
      <SimpleGrid>
        <Card
          imageProps={{
            alt: "Alt text",
            aspectRatio: "twoByOne",
            src: "//placekitten.com/500/200",
          }}
        >
          <CardHeading level="h2">Card Heading</CardHeading>
          <CardContent>
            Vestibulum id ligula porta felis euismod semper. Nulla vitae elit
            libero, a pharetra augue.
          </CardContent>
        </Card>
        <Card
          imageProps={{
            alt: "Alt text",
            aspectRatio: "twoByOne",
            src: "//placekitten.com/400/220",
          }}
        >
          <CardHeading level="h3">Card Heading</CardHeading>
          <CardContent>
            Vestibulum id ligula porta felis euismod semper. Nulla vitae elit
            libero, a pharetra augue.
          </CardContent>
        </Card>
        <Card
          imageProps={{
            alt: "Alt text",
            aspectRatio: "twoByOne",
            src: "//placekitten.com/400/240",
          }}
        >
          <CardHeading level="h3">Card Heading</CardHeading>
          <CardContent>
            Vestibulum id ligula porta felis euismod semper. Nulla vitae elit
            libero, a pharetra augue.
          </CardContent>
        </Card>
      </SimpleGrid>
    );

    expect(await axe(container)).toHaveNoViolations();
  });
});

describe("SimpleGrid", () => {
  it("Renders the SimpleGrid UI snapshot correctly", () => {
    const tree = renderer.create(<SimpleGrid id="test-grid" />).toJSON();
    const withChakraProps = renderer
      .create(<SimpleGrid id="chakra" p="20px" color="ui.error.primary" />)
      .toJSON();
    const withOtherProps = renderer
      .create(<SimpleGrid id="props" data-testid="props" />)
      .toJSON();

    expect(tree).toMatchSnapshot();
    expect(withChakraProps).toMatchSnapshot();
    expect(withOtherProps).toMatchSnapshot();
  });

  it("Renders SimpleGrid component", () => {
    const utils = render(<SimpleGrid id="test-grid-render" />);
    expect(
      utils.container.querySelector("#test-grid-render")
    ).toBeInTheDocument();
  });

  it("passes a ref to the div element", () => {
    const ref = React.createRef<HTMLDivElement>();
    const { container } = render(
      <SimpleGrid id="grid-ref" ref={ref}>
        <p>test1</p>
        <p>test2</p>
      </SimpleGrid>
    );

    expect(container.querySelector("div")).toBe(ref.current);
  });
});
