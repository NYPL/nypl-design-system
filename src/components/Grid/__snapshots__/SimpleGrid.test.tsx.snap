// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SimpleGrid Renders the SimpleGrid UI snapshot correctly 1`] = `
<div
  className="css-1pbowfz"
  id="test-grid"
/>
`;
